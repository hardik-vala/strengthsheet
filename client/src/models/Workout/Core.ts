export enum ExerciseUnit {
  kilometers,
  lb,
  meters,
  miles,
  millimeters,
  reps,
  time
}

export enum SetType {
  Drop,
  Failure,
  Warmup,
  Working,
}

export interface Circuit {
  key: string;
  exercises: { exercise: Exercise, note: string | null }[];
}

export interface Exercise {
  key: string;
  displayName: string;
  measures: ExerciseMeasure[];
}

export interface ExerciseMeasure {
  key: string;
  displayName: string;
  unit: ExerciseUnit;
}

export interface DrillSet {
  index: number;
  setType: SetType;
}

export class WorkoutValueKey {

  readonly circuitKey: string | null;
  readonly exerciseKey: string;
  readonly setIndex: number;
  readonly setType: SetType;
  readonly measureKey: string;

  private constructor(circuitKey: string | null, exerciseKey: string, setIndex: number, setType: SetType, measureKey: string) {
    this.circuitKey = circuitKey;
    this.exerciseKey = exerciseKey;
    this.setIndex = setIndex;
    this.setType = setType;
    this.measureKey = measureKey;
  }

  private freeze() {
    Object.freeze(this);
  }

  public static create(circuitKey: string | null, exerciseKey: string, setIndex: number, setType: SetType, measureKey: string): WorkoutValueKey {
    if (circuitKey) {
      return WorkoutValueKey.createFromCircuit(circuitKey, exerciseKey, setIndex, setType, measureKey);
    }

    return WorkoutValueKey.createFromExercise(exerciseKey, setIndex, setType, measureKey);
  }

  public static createFromExercise(exerciseKey: string, setIndex: number, setType: SetType, measureKey: string): WorkoutValueKey {
    return WorkoutValueKey.createFromCircuit(null, exerciseKey, setIndex, setType, measureKey);
  }

  public static createFromCircuit(circuitKey: string, exerciseKey: string, setIndex: number, setType: SetType, measureKey: string): WorkoutValueKey {
    const key = new WorkoutValueKey(circuitKey, exerciseKey, setIndex, setType, measureKey);
    key.freeze();
    return key;
  }

  public equals(other: WorkoutValueKey) {
    return this.circuitKey === other.circuitKey
      && this.exerciseKey === other.exerciseKey
      && this.setIndex === other.setIndex
      && this.setType === other.setType
      && this.measureKey === other.measureKey
  }

  public toString(): string {
    let s = `${this.exerciseKey}:${this.setIndex}:${this.setType.toString()}:${this.measureKey}`;

    if (this.circuitKey) {
      s = `${this.circuitKey}:${s}`
    }

    return s;
  }
}

const TIME_MINUTES_SECONDS_REGEX = /^[0-5][0-9]:[0-5][0-9]$/;
const TIME_HOURS_MINUTES_SECONDS_REGEX = /^[0-9][0-9]:[0-5][0-9]:[0-5][0-9]$/;

export function isValidMeasureValue(value: string, unit: ExerciseUnit): boolean {
  value = value.trim();

  if (value.length === 0) {
    return false;
  }

  switch (unit) {
    case ExerciseUnit.kilometers:
    case ExerciseUnit.lb:
    case ExerciseUnit.meters:
    case ExerciseUnit.miles:
    case ExerciseUnit.millimeters:
      const parsedFloat = Number(value);
      return Number.isFinite(parsedFloat) && parsedFloat > 0;
    case ExerciseUnit.reps:
      const parsedInt = Number(value);
      return Number.isInteger(parsedInt) && parsedInt > 0;
    case ExerciseUnit.time:
      return isValidTime(value);
  }
}

function isValidTime(time: string): boolean {
  return (
    TIME_MINUTES_SECONDS_REGEX.test(time) ||
    TIME_HOURS_MINUTES_SECONDS_REGEX.test(time)
  );
}

export function getSetTypeDisplayName(setType: SetType): string {
  switch(setType) {
    case SetType.Drop:
      return "Drop";
    case SetType.Failure:
      return "Failure";
    case SetType.Warmup:
      return "Warmup";
    case SetType.Working:
      return "Working";
    default:
      throw new Error(`Unrecognized set type: ${setType}`);
  }
}