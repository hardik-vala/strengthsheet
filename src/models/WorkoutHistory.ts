import { WorkoutHistoryRecord } from "./WorkoutHistoryRecord";

export type WorkoutHistory = WorkoutHistoryRecord[];