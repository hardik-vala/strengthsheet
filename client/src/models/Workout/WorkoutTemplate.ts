import { Circuit, DrillSet, Exercise } from "./Core";

export interface WorkoutTemplate {
	key: string;
	displayName: string;
	iconKey: string;
	note: string;
	drills: DrillTemplate[];	
}

export type DrillTemplate = ExerciseTemplate | CircuitTemplate;

export interface ExerciseTemplate {
	exercise: Exercise;
	displayName: string | null;
	iconKey: string;
	note: string | null;
	sets: DrillSet[];
}

export interface CircuitTemplate {
	circuit: Circuit;
	displayName: string | null;
	iconKey: string;
	note: string | null;
	sets: DrillSet[];
}
