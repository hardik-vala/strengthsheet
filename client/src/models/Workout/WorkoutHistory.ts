import { WorkoutValueKey } from "./Core";
import { WorkoutTemplate } from "./WorkoutTemplate";

export interface WorkoutHistory {
  workoutTemplate: WorkoutTemplate
  records: WorkoutHistoryRecord[];
}

export interface WorkoutHistoryRecord {
  startTimestamp: Date;
  elapsedTime: number;
  exercises: ExerciseHistoryRecord[];
};

export interface ExerciseHistoryRecord {
  key: WorkoutValueKey;
  value: string;
}
