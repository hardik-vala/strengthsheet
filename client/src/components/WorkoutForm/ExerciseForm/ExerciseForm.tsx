import { View } from "react-native";
import { Text } from "react-native-paper";
import { WorkoutHistory } from "../../../models/Workout/WorkoutHistory";
import { ExerciseTemplate } from "../../../models/Workout/WorkoutTemplate";
import { ExerciseFormBody } from "../ExerciseFormBody/ExerciseFormBody";
import { WorkoutValues } from "../common";
import { styles } from "../style";

export interface ExerciseFormProps {
  exerciseTemplate: ExerciseTemplate;
  workoutHistory: WorkoutHistory | null;
  workoutValues: WorkoutValues;
  onUpdateWorkoutValues: (updatedWorkoutValues: WorkoutValues) => void;
}

export function ExerciseForm({
  exerciseTemplate,
  workoutHistory,
  workoutValues,
  onUpdateWorkoutValues,
}: ExerciseFormProps) {
  return (
    <>
      <Text variant="titleMedium" style={styles.drillTitle}>
        {exerciseTemplate.displayName}
      </Text>
      <View style={styles.drillNoteContainer}>
        <Text variant="bodyMedium" style={styles.drillNoteText}>
          {exerciseTemplate.note}
        </Text>
      </View>
      <ExerciseFormBody 
        exercise={exerciseTemplate.exercise}
        sets={exerciseTemplate.sets}
        workoutHistory={workoutHistory}
        workoutValues={workoutValues}
        onUpdateWorkoutValues={onUpdateWorkoutValues}
      />
    </>
  );
}
