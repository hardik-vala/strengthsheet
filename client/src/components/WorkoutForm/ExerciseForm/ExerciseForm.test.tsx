import { render, screen } from "@testing-library/react-native";
import React from "react";
import {
  ExerciseUnit,
  SetType,
  WorkoutValueKey,
} from "../../../models/Workout/Core";
import { WorkoutHistory } from "../../../models/Workout/WorkoutHistory";

const TEST_EXERCISE = {
  key: "test_exercise",
  displayName: "Test Exercise",
  measures: [
    { key: "lb", displayName: "lb", unit: ExerciseUnit.lb },
    {
      key: "test_measure",
      displayName: "reps",
      unit: ExerciseUnit.reps,
    },
  ],
};

const TEST_TEMPLATE = {
  exercise: TEST_EXERCISE,
  displayName: "Test Exercise",
  iconKey: "",
  note: null,
  sets: [
    { index: 1, setType: SetType.Working },
    { index: 2, setType: SetType.Working },
  ],
};

const TEST_HISTORY: WorkoutHistory = {
  workoutTemplate: null,
  records: [
    {
      startTimestamp: new Date("2023-08-28"),
      elapsedTime: 3600000,
      exercises: [
        {
          key: WorkoutValueKey.createFromExercise(
            "test_exercise",
            1,
            SetType.Working,
            "test_measure"
          ),
          value: "13",
        },
        {
          key: WorkoutValueKey.createFromExercise(
            "test_exercise",
            2,
            SetType.Working,
            "test_measure"
          ),
          value: "13",
        },
      ],
    },
    {
      startTimestamp: new Date("2023-08-27"),
      elapsedTime: 3600000,
      exercises: [
        {
          key: WorkoutValueKey.createFromExercise(
            "test_exercise",
            1,
            SetType.Working,
            "test_measure"
          ),
          value: "12",
        },
        {
          key: WorkoutValueKey.createFromExercise(
            "test_exercise",
            2,
            SetType.Working,
            "test_measure"
          ),
          value: "12",
        },
      ],
    },
  ],
};

jest.mock("../../../services/backendService", () => {
  return {
    fetchWorkoutHistory: jest.fn(),
    storeRecordInWorkoutHistory: jest.fn(),
  };
});

import { ExerciseForm } from "./ExerciseForm";

describe("ExerciseForm", () => {
  it("renders correctly", () => {
    render(
      <ExerciseForm
        exerciseTemplate={TEST_TEMPLATE}
        workoutHistory={TEST_HISTORY}
        workoutValues={{}}
        onUpdateWorkoutValues={() => {}}
      />
    );

    expect(screen).toMatchSnapshot();
  });

  it("renders with note correctly", () => {
    render(
      <ExerciseForm
        exerciseTemplate={{...TEST_TEMPLATE, note: "This is a test note."}}
        workoutHistory={TEST_HISTORY}
        workoutValues={{}}
        onUpdateWorkoutValues={() => {}}
      />
    );

    expect(screen).toMatchSnapshot();
  });
});
