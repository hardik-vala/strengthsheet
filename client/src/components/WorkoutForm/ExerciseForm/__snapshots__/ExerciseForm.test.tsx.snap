// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExerciseForm renders correctly 1`] = `
[
  <Text
    style={
      [
        {
          "textAlign": "left",
        },
        {
          "color": "rgba(28, 27, 31, 1)",
          "writingDirection": "ltr",
        },
        [
          {
            "fontFamily": "System",
            "fontSize": 16,
            "fontWeight": "500",
            "letterSpacing": 0.15,
            "lineHeight": 24,
          },
          {
            "color": "deepskyblue",
            "fontWeight": "bold",
            "marginVertical": 10,
          },
        ],
      ]
    }
  >
    Test Exercise
  </Text>,
  <View
    style={
      {
        "backgroundColor": "gray",
        "padding": 5,
      }
    }
  >
    <Text
      style={
        [
          {
            "textAlign": "left",
          },
          {
            "color": "rgba(28, 27, 31, 1)",
            "writingDirection": "ltr",
          },
          [
            {
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "400",
              "letterSpacing": 0.25,
              "lineHeight": 20,
            },
            {
              "color": "black",
            },
          ],
        ]
      }
    />
  </View>,
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 10,
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "column",
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "writingDirection": "ltr",
            },
            [
              {
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "400",
                "letterSpacing": 0.25,
                "lineHeight": 20,
              },
              {
                "color": "gray",
                "fontWeight": "bold",
                "width": 90,
              },
            ],
          ]
        }
      >
        Set
      </Text>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "writingDirection": "ltr",
            },
            [
              {
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "400",
                "letterSpacing": 0.25,
                "lineHeight": 20,
              },
              {
                "color": "gray",
                "fontWeight": "bold",
                "width": 90,
              },
            ],
          ]
        }
      >
        lb
      </Text>
      <Text
        style={
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "writingDirection": "ltr",
            },
            [
              {
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "400",
                "letterSpacing": 0.25,
                "lineHeight": 20,
              },
              {
                "color": "gray",
                "fontWeight": "bold",
                "width": 90,
              },
            ],
          ]
        }
      >
        reps
      </Text>
    </View>
  </View>,
  <View
    style={
      {
        "width": "100%",
      }
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "rgba(202, 196, 208, 1)",
            "height": 0.5,
          },
          undefined,
          false,
          false,
          undefined,
        ]
      }
    />
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "space-between",
          "padding": 10,
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "flexDirection": "column",
            "justifyContent": "center",
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "rgba(232, 222, 248, 1)",
              "borderRadius": 8,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
            }
          }
          testID="chip-container-outer-layer"
        >
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "rgba(232, 222, 248, 1)",
                "borderColor": "rgba(121, 116, 126, 1)",
                "borderRadius": 8,
                "borderStyle": "solid",
                "borderWidth": 0,
                "flex": undefined,
                "flexDirection": "column",
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0,
                "shadowRadius": 0,
              }
            }
            testID="chip-container"
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": true,
                  "expanded": undefined,
                  "selected": false,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  {
                    "overflow": "hidden",
                  },
                  [
                    {
                      "borderRadius": 8,
                    },
                    {
                      "width": "100%",
                    },
                  ],
                ]
              }
              testID="chip"
            >
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "paddingLeft": 4,
                      "position": "relative",
                    },
                    {
                      "paddingLeft": 0,
                    },
                    {
                      "paddingRight": 0,
                    },
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  selectable={false}
                  style={
                    [
                      {
                        "textAlign": "left",
                      },
                      {
                        "color": "rgba(28, 27, 31, 1)",
                        "writingDirection": "ltr",
                      },
                      [
                        {
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        [
                          {
                            "marginVertical": 6,
                            "textAlignVertical": "center",
                          },
                          {
                            "color": "rgba(29, 25, 43, 1)",
                            "fontFamily": "System",
                            "fontSize": 14,
                            "fontWeight": "500",
                            "letterSpacing": 0.1,
                            "lineHeight": 20,
                          },
                          {
                            "marginLeft": 12,
                            "marginRight": 12,
                          },
                          {
                            "fontSize": 12,
                          },
                        ],
                      ],
                    ]
                  }
                >
                  1
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "flexDirection": "row",
          }
        }
      >
        <View>
          <View
            style={
              {
                "marginRight": 15,
                "width": 75,
              }
            }
          >
            <View
              pointerEvents="none"
              style={
                [
                  {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 6,
                  },
                  {
                    "backgroundColor": "rgba(232, 222, 248, 1)",
                    "borderColor": "transparent",
                    "borderRadius": 4,
                    "borderWidth": 1,
                  },
                  {
                    "borderRadius": 5,
                  },
                ]
              }
              testID="text-input-outline"
            />
            <View>
              <View
                style={
                  [
                    {
                      "paddingBottom": 0,
                    },
                    {
                      "minHeight": 30,
                      "paddingTop": 8,
                    },
                  ]
                }
              >
                <TextInput
                  cursorColor="transparent"
                  defaultValue=""
                  editable={true}
                  maxFontSizeMultiplier={1.5}
                  multiline={false}
                  onBlur={[Function]}
                  onChangeText={[Function]}
                  onFocus={[Function]}
                  placeholder=""
                  placeholderTextColor="rgba(73, 69, 79, 1)"
                  selectionColor="transparent"
                  style={
                    [
                      {
                        "margin": 0,
                        "zIndex": 1,
                      },
                      {
                        "height": 30,
                      },
                      {
                        "paddingBottom": 7,
                        "paddingTop": 7,
                      },
                      {
                        "color": "rgba(28, 27, 31, 1)",
                        "fontFamily": "System",
                        "fontSize": 16,
                        "fontWeight": undefined,
                        "letterSpacing": 0.15,
                        "lineHeight": undefined,
                        "paddingHorizontal": 16,
                        "textAlign": "right",
                        "textAlignVertical": "center",
                      },
                      false,
                      [
                        {},
                      ],
                      {
                        "fontSize": 12,
                        "fontWeight": "bold",
                        "textAlign": "center",
                      },
                    ]
                  }
                  testID="text-input-outlined"
                  underlineColorAndroid="transparent"
                  value=""
                />
              </View>
            </View>
          </View>
        </View>
        <View>
          <View
            style={
              {
                "marginRight": 15,
                "width": 75,
              }
            }
          >
            <View
              pointerEvents="none"
              style={
                [
                  {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 6,
                  },
                  {
                    "backgroundColor": "rgba(232, 222, 248, 1)",
                    "borderColor": "transparent",
                    "borderRadius": 4,
                    "borderWidth": 1,
                  },
                  {
                    "borderRadius": 5,
                  },
                ]
              }
              testID="text-input-outline"
            />
            <View>
              <View
                style={
                  [
                    {
                      "paddingBottom": 0,
                    },
                    {
                      "minHeight": 30,
                      "paddingTop": 8,
                    },
                  ]
                }
              >
                <TextInput
                  cursorColor="transparent"
                  defaultValue=""
                  editable={true}
                  maxFontSizeMultiplier={1.5}
                  multiline={false}
                  onBlur={[Function]}
                  onChangeText={[Function]}
                  onFocus={[Function]}
                  placeholder="13"
                  placeholderTextColor="rgba(73, 69, 79, 1)"
                  selectionColor="transparent"
                  style={
                    [
                      {
                        "margin": 0,
                        "zIndex": 1,
                      },
                      {
                        "height": 30,
                      },
                      {
                        "paddingBottom": 7,
                        "paddingTop": 7,
                      },
                      {
                        "color": "rgba(28, 27, 31, 1)",
                        "fontFamily": "System",
                        "fontSize": 16,
                        "fontWeight": undefined,
                        "letterSpacing": 0.15,
                        "lineHeight": undefined,
                        "paddingHorizontal": 16,
                        "textAlign": "right",
                        "textAlignVertical": "center",
                      },
                      false,
                      [
                        {},
                      ],
                      {
                        "fontSize": 12,
                        "fontWeight": "bold",
                        "textAlign": "center",
                      },
                    ]
                  }
                  testID="text-input-outlined"
                  underlineColorAndroid="transparent"
                  value=""
                />
              </View>
            </View>
          </View>
          <View>
            <View
              style={
                {
                  "backgroundColor": "rgba(255, 251, 254, 1)",
                }
              }
            >
              <View
                accessibilityRole="button"
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": false,
                    "expanded": false,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  [
                    {
                      "overflow": "hidden",
                    },
                    [
                      {
                        "paddingRight": 24,
                        "paddingVertical": 8,
                      },
                      {
                        "marginBottom": -10,
                        "marginLeft": 0,
                        "marginTop": -25,
                      },
                    ],
                  ]
                }
              >
                <View
                  pointerEvents="none"
                  style={
                    {
                      "flexDirection": "row",
                      "marginVertical": 6,
                    }
                  }
                >
                  <View
                    pointerEvents="box-none"
                    style={
                      [
                        {
                          "alignItems": "center",
                          "justifyContent": "center",
                        },
                        {
                          "alignSelf": "center",
                          "marginLeft": 16,
                          "marginRight": 0,
                          "marginVertical": 0,
                        },
                      ]
                    }
                  >
                    <Text />
                  </View>
                  <View
                    style={
                      [
                        {
                          "paddingLeft": 16,
                        },
                        {
                          "flex": 1,
                          "justifyContent": "center",
                        },
                      ]
                    }
                  >
                    <Text
                      numberOfLines={1}
                      selectable={false}
                      style={
                        [
                          {
                            "textAlign": "left",
                          },
                          {
                            "color": "rgba(28, 27, 31, 1)",
                            "fontFamily": "System",
                            "fontWeight": "400",
                            "letterSpacing": 0,
                          },
                          {
                            "writingDirection": "ltr",
                          },
                          [
                            {
                              "fontSize": 16,
                            },
                            {
                              "color": "rgba(28, 27, 31, 1)",
                            },
                            {
                              "fontSize": 12,
                              "marginLeft": -5,
                            },
                          ],
                        ]
                      }
                    >
                      History
                    </Text>
                  </View>
                  <View
                    style={
                      [
                        {
                          "marginVertical": 6,
                          "paddingLeft": 8,
                        },
                        undefined,
                      ]
                    }
                  >
                    <Text />
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>,
  <View
    style={
      {
        "width": "100%",
      }
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "rgba(202, 196, 208, 1)",
            "height": 0.5,
          },
          undefined,
          false,
          false,
          undefined,
        ]
      }
    />
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "space-between",
          "padding": 10,
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "flexDirection": "column",
            "justifyContent": "center",
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "rgba(232, 222, 248, 1)",
              "borderRadius": 8,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
            }
          }
          testID="chip-container-outer-layer"
        >
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "rgba(232, 222, 248, 1)",
                "borderColor": "rgba(121, 116, 126, 1)",
                "borderRadius": 8,
                "borderStyle": "solid",
                "borderWidth": 0,
                "flex": undefined,
                "flexDirection": "column",
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0,
                "shadowRadius": 0,
              }
            }
            testID="chip-container"
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": true,
                  "expanded": undefined,
                  "selected": false,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  {
                    "overflow": "hidden",
                  },
                  [
                    {
                      "borderRadius": 8,
                    },
                    {
                      "width": "100%",
                    },
                  ],
                ]
              }
              testID="chip"
            >
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "paddingLeft": 4,
                      "position": "relative",
                    },
                    {
                      "paddingLeft": 0,
                    },
                    {
                      "paddingRight": 0,
                    },
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  selectable={false}
                  style={
                    [
                      {
                        "textAlign": "left",
                      },
                      {
                        "color": "rgba(28, 27, 31, 1)",
                        "writingDirection": "ltr",
                      },
                      [
                        {
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        [
                          {
                            "marginVertical": 6,
                            "textAlignVertical": "center",
                          },
                          {
                            "color": "rgba(29, 25, 43, 1)",
                            "fontFamily": "System",
                            "fontSize": 14,
                            "fontWeight": "500",
                            "letterSpacing": 0.1,
                            "lineHeight": 20,
                          },
                          {
                            "marginLeft": 12,
                            "marginRight": 12,
                          },
                          {
                            "fontSize": 12,
                          },
                        ],
                      ],
                    ]
                  }
                >
                  2
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "flexDirection": "row",
          }
        }
      >
        <View>
          <View
            style={
              {
                "marginRight": 15,
                "width": 75,
              }
            }
          >
            <View
              pointerEvents="none"
              style={
                [
                  {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 6,
                  },
                  {
                    "backgroundColor": "rgba(232, 222, 248, 1)",
                    "borderColor": "transparent",
                    "borderRadius": 4,
                    "borderWidth": 1,
                  },
                  {
                    "borderRadius": 5,
                  },
                ]
              }
              testID="text-input-outline"
            />
            <View>
              <View
                style={
                  [
                    {
                      "paddingBottom": 0,
                    },
                    {
                      "minHeight": 30,
                      "paddingTop": 8,
                    },
                  ]
                }
              >
                <TextInput
                  cursorColor="transparent"
                  defaultValue=""
                  editable={true}
                  maxFontSizeMultiplier={1.5}
                  multiline={false}
                  onBlur={[Function]}
                  onChangeText={[Function]}
                  onFocus={[Function]}
                  placeholder=""
                  placeholderTextColor="rgba(73, 69, 79, 1)"
                  selectionColor="transparent"
                  style={
                    [
                      {
                        "margin": 0,
                        "zIndex": 1,
                      },
                      {
                        "height": 30,
                      },
                      {
                        "paddingBottom": 7,
                        "paddingTop": 7,
                      },
                      {
                        "color": "rgba(28, 27, 31, 1)",
                        "fontFamily": "System",
                        "fontSize": 16,
                        "fontWeight": undefined,
                        "letterSpacing": 0.15,
                        "lineHeight": undefined,
                        "paddingHorizontal": 16,
                        "textAlign": "right",
                        "textAlignVertical": "center",
                      },
                      false,
                      [
                        {},
                      ],
                      {
                        "fontSize": 12,
                        "fontWeight": "bold",
                        "textAlign": "center",
                      },
                    ]
                  }
                  testID="text-input-outlined"
                  underlineColorAndroid="transparent"
                  value=""
                />
              </View>
            </View>
          </View>
        </View>
        <View>
          <View
            style={
              {
                "marginRight": 15,
                "width": 75,
              }
            }
          >
            <View
              pointerEvents="none"
              style={
                [
                  {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 6,
                  },
                  {
                    "backgroundColor": "rgba(232, 222, 248, 1)",
                    "borderColor": "transparent",
                    "borderRadius": 4,
                    "borderWidth": 1,
                  },
                  {
                    "borderRadius": 5,
                  },
                ]
              }
              testID="text-input-outline"
            />
            <View>
              <View
                style={
                  [
                    {
                      "paddingBottom": 0,
                    },
                    {
                      "minHeight": 30,
                      "paddingTop": 8,
                    },
                  ]
                }
              >
                <TextInput
                  cursorColor="transparent"
                  defaultValue=""
                  editable={true}
                  maxFontSizeMultiplier={1.5}
                  multiline={false}
                  onBlur={[Function]}
                  onChangeText={[Function]}
                  onFocus={[Function]}
                  placeholder="13"
                  placeholderTextColor="rgba(73, 69, 79, 1)"
                  selectionColor="transparent"
                  style={
                    [
                      {
                        "margin": 0,
                        "zIndex": 1,
                      },
                      {
                        "height": 30,
                      },
                      {
                        "paddingBottom": 7,
                        "paddingTop": 7,
                      },
                      {
                        "color": "rgba(28, 27, 31, 1)",
                        "fontFamily": "System",
                        "fontSize": 16,
                        "fontWeight": undefined,
                        "letterSpacing": 0.15,
                        "lineHeight": undefined,
                        "paddingHorizontal": 16,
                        "textAlign": "right",
                        "textAlignVertical": "center",
                      },
                      false,
                      [
                        {},
                      ],
                      {
                        "fontSize": 12,
                        "fontWeight": "bold",
                        "textAlign": "center",
                      },
                    ]
                  }
                  testID="text-input-outlined"
                  underlineColorAndroid="transparent"
                  value=""
                />
              </View>
            </View>
          </View>
          <View>
            <View
              style={
                {
                  "backgroundColor": "rgba(255, 251, 254, 1)",
                }
              }
            >
              <View
                accessibilityRole="button"
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": false,
                    "expanded": false,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  [
                    {
                      "overflow": "hidden",
                    },
                    [
                      {
                        "paddingRight": 24,
                        "paddingVertical": 8,
                      },
                      {
                        "marginBottom": -10,
                        "marginLeft": 0,
                        "marginTop": -25,
                      },
                    ],
                  ]
                }
              >
                <View
                  pointerEvents="none"
                  style={
                    {
                      "flexDirection": "row",
                      "marginVertical": 6,
                    }
                  }
                >
                  <View
                    pointerEvents="box-none"
                    style={
                      [
                        {
                          "alignItems": "center",
                          "justifyContent": "center",
                        },
                        {
                          "alignSelf": "center",
                          "marginLeft": 16,
                          "marginRight": 0,
                          "marginVertical": 0,
                        },
                      ]
                    }
                  >
                    <Text />
                  </View>
                  <View
                    style={
                      [
                        {
                          "paddingLeft": 16,
                        },
                        {
                          "flex": 1,
                          "justifyContent": "center",
                        },
                      ]
                    }
                  >
                    <Text
                      numberOfLines={1}
                      selectable={false}
                      style={
                        [
                          {
                            "textAlign": "left",
                          },
                          {
                            "color": "rgba(28, 27, 31, 1)",
                            "fontFamily": "System",
                            "fontWeight": "400",
                            "letterSpacing": 0,
                          },
                          {
                            "writingDirection": "ltr",
                          },
                          [
                            {
                              "fontSize": 16,
                            },
                            {
                              "color": "rgba(28, 27, 31, 1)",
                            },
                            {
                              "fontSize": 12,
                              "marginLeft": -5,
                            },
                          ],
                        ]
                      }
                    >
                      History
                    </Text>
                  </View>
                  <View
                    style={
                      [
                        {
                          "marginVertical": 6,
                          "paddingLeft": 8,
                        },
                        undefined,
                      ]
                    }
                  >
                    <Text />
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>,
]
`;
